### ISSUE TITLE
Allow to define `vendored_dirs`, `data_only_dirs` and `subdir` rules from inside `dirs`

### ISSUE THREAD

**Issue by jchavarri** [ocaml/dune#6978]
I understand the original idea from dune was to place individual `dune` files in different folders to help reorganize code without changes in config. Over time, use cases to define rules that affect a folder from parent folders increasingly arrive: vendoring, consolidate configurations in a single place and such.

There are a few stanzas that relate to defining rules from a parent folder (excluding deprecated ones like `ignored_subdirs`): `dirs`, `vendored_dirs`, `data_only_dirs`, `subdir`.

Having these configurations arranged in different stanzas leads to more verbosity, and is more error prone. For example, all the stanzas above depend on `dirs`: the rules defined in `subdir` might be ignored if the dir being targeted is not included in `dirs`, which can be confusing.

I propose to consolidate these configurations under `dirs`, e.g.:

```lisp
(dirs
 (dir
  (name foo)
  (vendored true))
 (dir
  (name bar)
  (data_only true)
  (rule
   (with-stdout-to
    bar
    (echo baz)))))
```

This can be combined with `:standard` to refine the configuration of a given dir:

```
(dirs
 :standard
 (dir
  (name foo)
  (vendored true))
```

- `dirs` can take a list of strings or a list of `dir` fields.
- `vendored_dirs` stanza can be replaced with a `vendored` field inside the `dir`
- `data_only_dirs` stanza can be replaced with `data_only` field
- `subdir` stanza can be replaced with `rule` field

I believe this change could be added in a non-breaking way.

The downsides are:
- added complexity of maintaining 2 ways to express the same thing
- more verbosity for simple cases ("I just want to define a single vendored dir")

The first could be mitigated by deprecating the `vendored_dirs`, `data_only_dirs`, `subdir` stanzas at some point in the future. For the second one there is no fix I can think of.

What do you think?

**Comment by rgrinberg** [comment]
How would we apply configuration to groups of directories. Right now we can do

```
(data_only_dirs * \ special_dir)
```

To make everything data only except for `special_dir`.

**Comment by jchavarri** [comment]
Yeah, that configuration should be done on a per-dir basis. Unless it is possible to "push" the star operator down to the `name` field level? e.g.

```
(dirs
 (dir
  (name *)
  (data_only true))
 (dir
  (name special_dir)
  (vendored true)
  (data_only false)))
```

**Comment by jchavarri** [comment]
How would an example with multiple dirs like `*` and `special_dir` would look like with your proposal?
